src/lib/flow/sol-flow-js.c

```diff
@@ -0,0 +1,1422 @@
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <float.h>
+#include <stdio.h>
+
+#include "duktape.h"
+#include "sol-arena.h"
+#include "sol-flow-internal.h"
+#include "sol-flow-js.h"
+#include "sol-str-table.h"
+
+#define PORTS_IN_METHODS_LENGTH 3
+#define PORTS_IN_CONNECT_INDEX 0
+#define PORTS_IN_DISCONNECT_INDEX 1
+#define PORTS_IN_PROCESS_INDEX 2
+
+#define PORTS_OUT_METHODS_LENGTH 2
+#define PORTS_OUT_CONNECT_INDEX 0
+#define PORTS_OUT_DISCONNECT_INDEX 1
+
+/* Contains information specific to a type based on JS. */
+struct flow_js_type {
+    struct sol_flow_node_type base;
+
+    struct sol_vector ports_in;
+    struct sol_vector ports_out;
+
+    struct sol_arena *str_arena;
+
+    char *js_content_buf;
+    size_t js_content_buf_len;
+};
+
+struct flow_js_port_in {
+    struct sol_flow_port_type_in type;
+    char *name;
+    char *type_name;
+};
+
+struct flow_js_port_out {
+    struct sol_flow_port_type_out type;
+    char *name;
+    char *type_name;
+};
+
+/* Contains information specific to a node of a JS node type. */
+struct flow_js_data {
+    /* Each node keeps its own JavaScript context and global object. */
+    struct duk_context *duk_ctx;
+};
+
+static duk_ret_t
+send_boolean_packet(struct sol_flow_node *node, uint16_t port, duk_context *ctx)
+{
+    bool value = duk_require_boolean(ctx, 1);
+
+    sol_flow_send_boolean_packet(node, port, value);
```

Comment:
Do we want to ignore these errors? Shouldn't an exception be raised if these functions return <0?